### PR TITLE
Add compiler changes for enabling Error BP in `on fail` clause

### PR DESCRIPTION
**Pull request by suleka96**
## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 1]
```diff
@@ -3870,6 +3870,11 @@ public void visit(BLangOnFailClause onFailClause, AnalyzerData data) {
         SymbolEnv onFailEnv = SymbolEnv.createBlockEnv(onFailClause.body, data.env);
         VariableDefinitionNode onFailVarDefNode = onFailClause.variableDefinitionNode;
         if (onFailVarDefNode != null) {
+            BLangVariable variableNode = (BLangVariable) onFailVarDefNode.getVariable();
+            if (variableNode.getKind() != NodeKind.VARIABLE && variableNode.getKind() != NodeKind.ERROR_VARIABLE) {
+                dlog.error(variableNode.pos, DiagnosticErrorCode.INVALID_BINDING_PATTERN_IN_ON_FAIL);
+                return;
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:3876]: If we continue the analysis instead of returning we can log errors relevant to other binding patterns too. Not much value added here, but will help if they meant to make them part of detail binding patterns.

File: compiler/ballerina-lang/src/main/java/org/ballerinalang/util/diagnostic/DiagnosticErrorCode.java [hunk 2]
```diff
@@ -799,7 +799,9 @@ public enum DiagnosticErrorCode implements DiagnosticCode {
     EMPTY_REGEXP_STRING_DISALLOWED(
             "BCS4044", "empty.regexp.string.disallowed"),
     UNSUPPORTED_EMPTY_CHARACTER_CLASS(
-            "BCS4045", "unsupported.empty.character.class")
+            "BCS4045", "unsupported.empty.character.class"),
+
+    INVALID_BINDING_PATTERN_IN_ON_FAIL("BCS4046", "invalid.binding.pattern.in.on.fail")
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/ballerinalang/util/diagnostic/DiagnosticErrorCode.java:804]:
```suggestion
    INVALID_BINDING_PATTERN_IN_ON_FAIL("BCE4046", "invalid.binding.pattern.in.on.fail")
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 3]
```diff
@@ -5187,6 +5177,38 @@ onFailErrorVariableSymbol.name.value, onFailErrorVariableSymbol.type, rewrite(fa
         return fail;
     }
 
+    private void handleOnFailErrorVarDefNode(VariableDefinitionNode onFailVarDefNode, BLangBlockStmt onFailBody,
+                                             BLangFail fail) {
+        if (onFailVarDefNode != null) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:5182]: Check inverse and return early.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 4]
```diff
@@ -5187,6 +5177,38 @@ onFailErrorVariableSymbol.name.value, onFailErrorVariableSymbol.type, rewrite(fa
...
+            BLangStatement stmt;
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:5183]: Do you need this separate variable? Just add to statements instead of assigning to a temporary variable first?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 5]
```diff
@@ -5187,6 +5177,38 @@ onFailErrorVariableSymbol.name.value, onFailErrorVariableSymbol.type, rewrite(fa
...
+            if (variableNode.getKind() == NodeKind.ERROR_VARIABLE) {
+                stmt = handleErrorBPInOnFail((BLangErrorVariable) variableNode, fail);
+            } else {
+                stmt = handleCaptureBPInOnFail((BLangSimpleVariable) variableNode, fail);
+            }
+            onFailBody.stmts.add(0, stmt);
+        }
+    }
+
+    private BLangStatement handleErrorBPInOnFail (BLangErrorVariable varNode, BLangFail fail) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:5194]:
```suggestion
    private BLangStatement handleErrorBPInOnFail(BLangErrorVariable varNode, BLangFail fail) {
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 6]
```diff
@@ -5187,6 +5177,38 @@ onFailErrorVariableSymbol.name.value, onFailErrorVariableSymbol.type, rewrite(fa
...
+        BLangErrorVariable errorVar = ASTBuilderUtil.createErrorVariable(varNode.pos,
+                varNode.getBType(), rewrite(fail.expr, env), varNode.message, varNode.cause, varNode.restDetail,
+                varNode.detail);
+        BLangErrorVariableDef errorVarDef = ASTBuilderUtil.createErrorVariableDef(onFailClause.pos, errorVar);
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:5198]: Is the pos correct? Shouldn't it be that of the error BP?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java [hunk 7]
```diff
@@ -2269,9 +2269,12 @@ public void visit(BLangRecordVariableDef bLangRecordVariableDef) {
     public void visit(BLangErrorVariable bLangErrorVariable) {
         analyzeNode(bLangErrorVariable.typeNode, env);
         populateUnusedVariableMapForNonSimpleBindingPatternVariables(this.unusedLocalVariables, bLangErrorVariable);
-        this.currDependentSymbolDeque.push(bLangErrorVariable.symbol);
-        analyzeNode(bLangErrorVariable.expr, env);
-        this.currDependentSymbolDeque.pop();
+        // When we visit this path from on fail clause the symbol is null
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java:2272]: Is this true for capture binding patterns?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 8]
```diff
@@ -2106,12 +2106,12 @@ void handleDeclaredVarInForeach(BLangVariable variable, BType rhsType, SymbolEnv
                 break;
             case ERROR_VARIABLE:
                 BLangErrorVariable errorVariable = (BLangErrorVariable) variable;
+                errorVariable.setBType(rhsType);
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:2109]: Is this correct when the RHS type isn't an error type?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 9]
```diff
@@ -3869,28 +3876,71 @@ public void visit(BLangOnFailClause onFailClause, AnalyzerData data) {
...
+        onFailErrTypes.pop();
         data.env = onFailEnv;
         analyzeStmt(onFailClause.body, data);
     }
 
+    private void updateBreakMode(boolean onFailExists, BLangBlockStmt blockStmt, BLangOnFailClause onFailClause,
+                                 AnalyzerData data) {
+        if (onFailExists) {
+            blockStmt.failureBreakMode = getPossibleBreakMode(!data.commonAnalyzerData.errorTypes.peek().isEmpty());
+            this.analyzeNode(onFailClause, data);
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:3926]: Can we rename this since it doesn't just update break mode, but also analyzes the node?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 10]
```diff
@@ -3869,28 +3876,71 @@ public void visit(BLangOnFailClause onFailClause, AnalyzerData data) {
...
         }
+
+        onFailErrTypes.pop();
         data.env = onFailEnv;
         analyzeStmt(onFailClause.body, data);
     }
 
+    private void updateBreakMode(boolean onFailExists, BLangBlockStmt blockStmt, BLangOnFailClause onFailClause,
+                                 AnalyzerData data) {
+        if (onFailExists) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:3924]: Check inverse and return early.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 11]
```diff
@@ -3869,28 +3876,71 @@ public void visit(BLangOnFailClause onFailClause, AnalyzerData data) {
...
+
+            for (BType errorType : onFailErrTypes.peek()) {
+                if (onFailVarNode.getBType() != null && !types.isAssignable(errorType, onFailVarNode.getBType())) {
+                    dlog.error(onFailVarNode.pos, DiagnosticErrorCode.INCOMPATIBLE_ON_FAIL_ERROR_DEFINITION, errorType,
+                            onFailVarNode.getBType());
+                }
+            }
         }
+
+        onFailErrTypes.pop();
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:3917]: This should ideally happen in the methods where you push to `onFailErrTypes` IMO.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java [hunk 12]
```diff
@@ -2269,9 +2269,12 @@ public void visit(BLangRecordVariableDef bLangRecordVariableDef) {
...
+        if (bLangErrorVariable.symbol != null) {
+            this.currDependentSymbolDeque.push(bLangErrorVariable.symbol);
+            analyzeNode(bLangErrorVariable.expr, env);
+            this.currDependentSymbolDeque.pop();
+        }
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java:2277]: Unused variable errors are not logged for something like

```ballerina
public function main() {
    do {

    } on fail Error error(i = i) {

    }
}

type Error error<record {| int i; |}>;
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 13]
```diff
@@ -3869,28 +3876,71 @@ public void visit(BLangOnFailClause onFailClause, AnalyzerData data) {
...
+            BType failErrorType;
+            LinkedHashSet<BType> onFailErrType = onFailErrTypes.peek();
+            boolean isVarTypeHardcoded = false;
+            if (onFailErrType.size() == 1) {
+                failErrorType = onFailErrType.iterator().next();
+            } else if (onFailErrType.size() > 1) {
+                failErrorType = BUnionType.create(null, onFailErrType);
+            } else {
+                failErrorType = symTable.errorType;
+                isVarTypeHardcoded = true;
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:3896]: Why do we need to special case like this? Can't we use something like noType as the type instead to handle this? Or never.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 14]
```diff
@@ -3869,28 +3876,71 @@ public void visit(BLangOnFailClause onFailClause, AnalyzerData data) {
...
+                    onFailClause.isDeclaredWithVar, isVarTypeHardcoded, onFailEnv);
             BLangVariable onFailVarNode = (BLangVariable) onFailVarDefNode.getVariable();
-            if (!types.isAssignable(onFailVarNode.getBType(), symTable.errorType)) {
+
+            if (onFailVarNode.getBType() != null && !types.isAssignable(onFailVarNode.getBType(), symTable.errorType)) {
                 dlog.error(onFailVarNode.pos, DiagnosticErrorCode.INVALID_TYPE_DEFINITION_FOR_ERROR_VAR,
                         onFailVarNode.getBType());
             }
+
+            for (BType errorType : onFailErrTypes.peek()) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:3909]: Doesn't `handleForeachDefinitionVariables` already log errors when incompatible?

The following results in 3 errors.

```ballerina
public function main() {
    do {
        if true {
            fail error E1("e1", j = 1);
        }
        fail error E2("e2", k = 2);
    } on fail Error error(i = i) {

    }
}

type Error error<record {| int i; |}>;
type E1 error<record {| int j; |}>;
type E2 error<record {| int k; |}>;
```

```
ERROR [x.bal:(7:15,7:33)] incompatible error definition type: 'E1' will not be matched to 'Error'
ERROR [x.bal:(7:15,7:33)] incompatible error definition type: 'E2' will not be matched to 'Error'
ERROR [x.bal:(7:15,7:20)] incompatible types: expected '(E1|E2)', found 'Error'
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 15]
```diff
@@ -253,6 +253,40 @@ public boolean typeIncompatible(Location pos, BType actualType, BType expType) {
         return checkType(pos, actualType, expType, DiagnosticErrorCode.INCOMPATIBLE_TYPES) == symTable.semanticError;
     }
 
+    public BType getErrorTypes(BType bType) {
+        if (bType == null) {
+            return symTable.semanticError;
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:258]: codecov warning?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 16]
```diff
@@ -253,6 +253,40 @@ public boolean typeIncompatible(Location pos, BType actualType, BType expType) {
...
+        }
+
+        BType errorType = symTable.semanticError;
+
+        int tag = bType.tag;
+        if (tag == TypeTags.TYPEREFDESC) {
+            return getErrorTypes(Types.getReferredType(bType));
+        }
+        if (tag == TypeTags.ERROR) {
+            errorType = bType;
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:268]: Can just return? If-else blocks can be refactored with early returns.

File: langlib/langlib-test/src/test/java/org/ballerinalang/langlib/test/statements/foreach/ForeachNegativeTests.java [hunk 17]
```diff
@@ -104,6 +104,8 @@ public void testForeachSemanticsNegative() {
                 "mismatch with member type count", 342, 13);
         BAssertUtil.validateError(compile, index++, "invalid list binding pattern; member variable count " +
                 "mismatch with member type count", 346, 18);
+        BAssertUtil.validateError(compile, index++, "invalid error variable; expecting an error type " +
```
Comment (MaryamZi) [inline langlib/langlib-test/src/test/java/org/ballerinalang/langlib/test/statements/foreach/ForeachNegativeTests.java:107]: Did we create an issue to track allowing compatible binding patterns with unions?

File: langlib/langlib-test/src/test/resources/test-src/statements/foreach/foreach-on-fail.bal [hunk 18]
```diff
@@ -93,3 +93,412 @@ function testNestedForeachLoopBreak() returns string {
...
+    } on fail error error(msg, code = errCode, reason = errReason) {
+        testErrorCode = errCode;
+        testErrorReason = errReason;
+        testMessage = msg;
+    }
+    assertEquality(testMessage, "error!");
+    assertEquality(testErrorCode, 20);
+    assertEquality(testErrorReason, "deadlock condition");
+    assertEquality(testErrorCode is int, true);
+    assertEquality(testErrorReason is string, true);
```
Comment (MaryamZi) [inline langlib/langlib-test/src/test/resources/test-src/statements/foreach/foreach-on-fail.bal:138]: I wouldn't use module-level variables, won't allow us to properly run these tests concurrently.

How about asserting withing the on fail block and returning there. Then outside we can have something like an assert fail to make sure control was transferred to the on fail block.

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/retrystmt/retry_on_fail_negative_unreachable.bal [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+      fail error ErrorTypeA(TYPE_A_ERROR_REASON, message = "Error Type A");
+   }
+   on fail error e1 {
+       str += " -> Error caught !";
+   }
+   str += "-> Execution continues...";
+   return str;
+}
+
+function testOnFailWithUnion () returns string {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/retrystmt/retry_on_fail_negative_unreachable.bal:73]:
```suggestion
function testOnFailWithUnion() returns string {
```

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/statements/retrystmt/RetryStmtWithOnFailTest.java [hunk 20]
```diff
@@ -24,19 +24,22 @@
...
 
-    private CompileResult programFile, negativeFile;
+    private CompileResult programFile, negativeFile1, negativeFile2;
 
     @BeforeClass
     public void setup() {
         programFile = BCompileUtil.compile("test-src/statements/retrystmt/retry_stmt_with_on_fail.bal");
-        negativeFile = BCompileUtil.compile("test-src/statements/retrystmt/retry_on_fail_negative.bal");
+        negativeFile1 = BCompileUtil.compile("test-src/statements/retrystmt/retry_on_fail_negative.bal");
+        negativeFile2 = BCompileUtil.compile("test-src/statements/retrystmt/retry_on_fail_negative_unreachable.bal");
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/statements/retrystmt/RetryStmtWithOnFailTest.java:41]: Used only in one test method, right? Use a local variable instead.

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/retrystmt/retry_on_fail_negative_unreachable.bal
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/retrystmt/retry_on_fail_negative_unreachable.bal:1]: Add licence header.

How is this file related to your changes?

Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/resources/test-src/statements/retrystmt/retry_on_fail_negative_unreachable.bal:1]: IMO, the tests are overly complex for the errors we are trying to test. I understand that these are similar to existing tests, but not the easiest to read.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java (hunk 7 above)
Comment (suleka96) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java:2272]: for `on fail error msg` kind of scenarios the flow doesn't go into this method right

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java (hunk 8 above)
Comment (suleka96) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:2109]: We won't go into this code block if RHS is not assignable to type of the variable node, which is a `BErrorType` in this case.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java (hunk 11 above)
Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:3917]: Is this correct? Don't we push to `data.commonAnalyzerData.errorTypes` in the relevant statements? Errors are pushed to `data.commonAnalyzerData.errorTypes.peek()`?

File: langlib/langlib-test/src/test/java/org/ballerinalang/langlib/test/statements/foreach/ForeachNegativeTests.java (hunk 17 above)
Comment (suleka96) [reply langlib/langlib-test/src/test/java/org/ballerinalang/langlib/test/statements/foreach/ForeachNegativeTests.java:107]: ya https://github.com/ballerina-platform/ballerina-lang/issues/40377

**Comment by MaryamZi** [comment]
Review suggestions will be addressed separately.

Need to also fix

```ballerina
type ErrorTypeA distinct error;

type ErrorTypeB distinct error;

function fn() {
    transaction {
        var res = checkpanic commit;
        fail error ErrorTypeA("oops!", message = "Type A");
    } on fail ErrorTypeA e {

    }
}
```

Fails with

```ballerina
ERROR [temp.bal:(9:15,9:27)] incompatible error definition type: 'error' will not be matched to 'ErrorTypeA'
ERROR [temp.bal:(9:15,9:25)] incompatible types: expected '(error|ErrorTypeA)', found 'ErrorTypeA'
```
